### ISSUE TITLE
NPE when running embedded JMXTrans. Pool is null.

### ISSUE THREAD

**Issue by imochurad** [jmxtrans/jmxtrans#519]
I would like to embed JMX trans in my application code, I was following one of the examples that would fetch data and push it into Graphite or send it to standard output. I dumbed it down to the bare minimum. So that I poll data only from one Mbean and want to display it in console. Here is the code:

```
public class Main {

    private static final JsonPrinter printer = new JsonPrinter(System.out);
    
    public static void main(String[] args) throws Exception {
        
        
        Server.Builder serverBuilder = Server.builder()
                .setHost("localhost")
                .setPort("7199")
                .setNumQueryThreads(2);
        
        StdOutWriter sw = new StdOutWriter(ImmutableList.<String>of(), false, false, Collections.<String, Object>emptyMap());
        
        Query q4 = Query.builder()
                .setObj("java.lang:name=ParNew,type=GarbageCollector")
                .addAttr("LastGcInfo")
                .addOutputWriterFactories(sw)
                .build();
        serverBuilder.addQuery(q4);
        
        JmxProcess process = new JmxProcess(serverBuilder.build());
        printer.prettyPrint(process);
        Injector injector = JmxTransModule.createInjector(new JmxTransConfiguration());
        JmxTransformer transformer = injector.getInstance(JmxTransformer.class);
        transformer.executeStandalone(process);
    }

}
```

After running for few seconds I see following exception:

```
[jmxtrans-query-1] ERROR com.googlecode.jmxtrans.jmx.ProcessQueryThread - Error executing query Query(objectName=java.lang:name=ParNew,type=GarbageCollector, keys=[], attr=[LastGcInfo], typeNames=[], resultAlias=null, useObjDomainAsKey=false, allowDottedKeys=false, useAllTypeNames=false, outputWriterInstances=[]) on server Server(pid=null, host=localhost, port=7199, url=service:jmx:rmi:///jndi/rmi://localhost:7199/jmxrmi, cronExpression=null, numQueryThreads=2)
java.lang.NullPointerException
	at com.googlecode.jmxtrans.model.Server.execute(Server.java:261)
	at com.googlecode.jmxtrans.jmx.ProcessQueryThread.run(ProcessQueryThread.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
```

I had also stepped in debug mode, it is pool object that is null. What am I doing wrong? How do I properly initialize it?

**Comment by gehel** [comment]
You can set the pool of the `Server` with its builder ([`serverBuilder.setPool()`](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Server.java#L470)). The [`JmxTransModule`](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/guice/JmxTransModule.java#L81-L88) that you already use does provide initialisation of a pool. It is usually automatically injected when parsing json files, by [using the Jackson GuavaModule](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/guice/JmxTransModule.java#L174-L175).

Side note: if you want to embed jmxtrans into your application, you should probably have a look at [embeded-jmxtrans](https://github.com/jmxtrans/embedded-jmxtrans). Jmxtrans itself is not meant to be embedded. It has no public interface except its configuration files. Any internal class can (and will) change in future versions, with no warning whatsoever.

**Comment by imochurad** [comment]
What I need to do is collect the data from several JVMs deployed in the cloud environment. I cannot hard code host and url related info in the json file. I thought I might create a java worker application that would launch, discover JMX location-related info and then start jmxtrans process(es). I have initially looked at embedded-jmxtrans, it does have integration with Spring and servlets. What I need a simple class with exposed constructor, where I can pass all the configuration that I dynamically compose. What is the best approach for it?

**Comment by imochurad** [comment]
According to the slides from this presentation here (slide 27)
http://www.slideshare.net/cyrille.leclerc/open-source-monitoring-for-java-with-jmx-and-graphite
standalone jmxtrans is more suitable to what I am trying to achieve since it can connect to several remote jvms at once. Can this be done also with embedded jmxtrans? If not, I need to incorporate standalone jmxtrans. What would be the best practice in that scenario?

**Comment by gehel** [comment]
This seems to be fairly similar to a question that was asked recently on the google group. See [my answer there](https://groups.google.com/d/msg/jmxtrans/SUG3zCjfddM/V5taap8dAQAJ). In short, the stable API to jmxtrans is its configuration. So you should probably have an external process discovering your JVMs and writing the jmxtrans configuration. Jmxtrans should auto restart when is sees changes in config.

A lot of people use a similar approach with a configuration management tool (puppet, chef, ansible, ...) doing the orchestration, and jmxtrans just reading generated configuration.

**Comment by imochurad** [comment]
Do you know cases running standalone jmxtrans in Heroku with config self discovery? An example would be great. Every process in Heroku (web or worker) runs within a dyno that has an ephemeral file system. Even if I deploy standalone jmxtrans jar with an external config, I won't be able to update that config file from the discovery application since it will be launched on a different dyno. The only thing possible from config discovery app would be to update config vars for a dyno running jmxtrans via Heroku API... On a sidenote: how do you restart jmxtrans jar if there are changes to config?

**Comment by gehel** [comment]
You might want to have a look at [jmxtrans-agent](https://github.com/jmxtrans/jmxtrans-agent). This would allow you to run jmxtrans in the same jvm as the application under monitoring. This way, no discovery needs to happen.

Jmxtrans watches the config files and reload them if there are any changes. You can always restart it manually if you want to. The starting point to dig into how restart actually works is [here](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/JmxTransformer.java#L288-L299).

**Comment by imochurad** [comment]
From what I understand, jmxtrans-agent instruments only local JMX. It doesn't allow to do remote instrumentation. And that's not gonna work in Heroku since some of the services like Cassandra, Kafka, etc are available through 3rd party providers and we do not have access to the nodes, so I don't see how jmxtrans-agent can fit.
Initially, I was inclined to use jmxtrans API mode: 
https://github.com/jmxtrans/jmxtrans/wiki#api-mode 
So that I could discover first the infrastructure I have running in cloud, and then create the object model by hand and execute it. That would allow to achieve desired results.
That section references example classes. I took a sample code from https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-examples/src/main/java/com/googlecode/jmxtrans/example/InterestingInfo.java
But when I launched it, I got NPE that I described above. So that NPE set me off... Looks like there are missing pieces to the puzzle.

**Comment by gehel** [comment]
That section about API is a bit outdated. It should at least precise that there are no commitment done to keep this API stable.

Are you sure that you can access those 3rd party services via JMX in the first place? I have mostly no experience with Heroku, so that might be possible, but it would surprise me that 3rd party allow JMX access.

All that being said, feel free to extend jmxtrans with whatever code you might need. Or to embed it. Just be aware that things are subject to changes and this is clearly uncharted territory. [Here be dragons](https://en.wikipedia.org/wiki/Here_be_dragons).

**Comment by imochurad** [comment]
Yeah, 3rd parties in Heroku expose JMX urls on demand. And I was able to connect to remote JMX from my test app. I was doing it via j`avax.management.MBeanServerConnection` and all standard goodness. The only thing that I had to do extra was to import ssl cert to avoid ssl handshake exception.
Thanks for your replies. I might share my findings if that all works out fine.

**Comment by imochurad** [comment]
I am currently generating my json config file on the fly and here how it looks like (I am deliberately switching from GraphiteWriterFactory to GraphiteWriter):

`[{
	"host": "localhost",
	"port": 7199,
	"numQueryThreads": 1,
	"outputWriters": [{
		"@class": "com.googlecode.jmxtrans.model.output.GraphiteWriter",
		"rootPrefix": "289c4cd3-d317-4a38-848b-14281a433bae",
		"host": "localhost",
		"port": 2003
	}],
	"queries": [{
		"obj": "org.apache.cassandra.metrics:type=ClientRequest,scope=Write,name=TotalLatency",
		"attr": ["Count"],
		"resultAlias": "LatencyWrites",
		"allowDottedKeys": false,
		"useAllTypeNames": false,
		"useObjDomainAsKey": false
	}],
	"runPeriodSeconds": 5
}, {
	"host": "localhost",
	"port": 7199,
	"numQueryThreads": 2,
	"outputWriters": [{
		"@class": "com.googlecode.jmxtrans.model.output.GraphiteWriter",
		"rootPrefix": "289c4cd3-d317-4a38-848b-14281a433bae",
		"host": "localhost",
		"port": 2003
	}],
	"queries": [{
		"obj": "org.apache.cassandra.metrics:type=ClientRequest,scope=Write,name=Latency",
		"attr": ["OneMinuteRate"],
		"resultAlias": "ThroughputWrites",
		"allowDottedKeys": false,
		"useAllTypeNames": false,
		"useObjDomainAsKey": false
	}, {
		"obj": "org.apache.cassandra.metrics:type=ClientRequest,scope=Read,name=Latency",
		"attr": ["OneMinuteRate"],
		"resultAlias": "ThroughputReads",
		"allowDottedKeys": false,
		"useAllTypeNames": false,
		"useObjDomainAsKey": false
	}],
	"runPeriodSeconds": 10
}]`

Here is the code that generates transformer object:

```
    private static JmxTransformer createTransformer() {
        File f = new File(JMXTRANS_CONFIG_FILE);
        JmxTransConfiguration config = new JmxTransConfiguration();
        config.setJsonFile(f);
        Injector injector = JmxTransModule.createInjector(config);
        JmxTransformer transformer = injector.getInstance(JmxTransformer.class);
        return transformer;
    }
```

I get NPE:

```
13:43:50.923 [jmxtrans-result-7] WARN  c.g.jmxtrans.jmx.ResultProcessor - Could not write results [Result(attributeName=Count, className=com.yammer.metrics.reporting.JmxReporter$Counter, objDomain=org.apache.cassandra.metrics, typeName=type=ClientRequest,scope=Write,name=TotalLatency, values={Count=0}, epoch=1480527798057, keyAlias=LatencyWrites)] of query Query(objectName=org.apache.cassandra.metrics:type=ClientRequest,scope=Write,name=TotalLatency, keys=[], attr=[Count], typeNames=[], resultAlias=LatencyWrites, useObjDomainAsKey=false, allowDottedKeys=false, useAllTypeNames=false, outputWriterInstances=[]) to output writer GraphiteWriter(pool=org.apache.commons.pool.impl.GenericKeyedObjectPool@39c11e6c, rootPrefix=289c4cd3-d317-4a38-848b-14281a433bae, address=localhost/127.0.0.1:2003, onlyOnceLogger=com.googlecode.jmxtrans.util.OnlyOnceLogger@329a1243)
java.lang.NullPointerException: null
	at org.apache.commons.pool.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:1220)
	at com.googlecode.jmxtrans.model.output.GraphiteWriter.internalWrite(GraphiteWriter.java:116)
	at com.googlecode.jmxtrans.model.output.BaseOutputWriter.doWrite(BaseOutputWriter.java:157)
	at com.googlecode.jmxtrans.jmx.ResultProcessor$1.run(ResultProcessor.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)


13:43:51.904 [jmxtrans-result-7] WARN  c.g.jmxtrans.jmx.ResultProcessor - Could not write results [Result(attributeName=OneMinuteRate, className=com.yammer.metrics.reporting.JmxReporter$Timer, objDomain=org.apache.cassandra.metrics, typeName=type=ClientRequest,scope=Write,name=Latency, values={OneMinuteRate=0.0}, epoch=1480527829904, keyAlias=ThroughputWrites)] of query Query(objectName=org.apache.cassandra.metrics:type=ClientRequest,scope=Write,name=Latency, keys=[], attr=[OneMinuteRate], typeNames=[], resultAlias=ThroughputWrites, useObjDomainAsKey=false, allowDottedKeys=false, useAllTypeNames=false, outputWriterInstances=[]) to output writer GraphiteWriter(pool=org.apache.commons.pool.impl.GenericKeyedObjectPool@3070ff69, rootPrefix=289c4cd3-d317-4a38-848b-14281a433bae, address=localhost/127.0.0.1:2003, onlyOnceLogger=com.googlecode.jmxtrans.util.OnlyOnceLogger@ecf9fb3)
java.lang.NullPointerException: null
	at org.apache.commons.pool.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:1220)
	at com.googlecode.jmxtrans.model.output.GraphiteWriter.internalWrite(GraphiteWriter.java:116)
	at com.googlecode.jmxtrans.model.output.BaseOutputWriter.doWrite(BaseOutputWriter.java:157)
	at com.googlecode.jmxtrans.jmx.ResultProcessor$1.run(ResultProcessor.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

How do I fix it?

**Comment by gehel** [comment]
I'm probably missing a bit of context here. Is your code available somewhere public? Can I have a look?

I'll have a try anyway, to at least put you on the right way.

The object pool is injected in the [GraphiteWriter](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/GraphiteWriter.java#L156L159) with the help of the [Guice GuavaModule](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/guice/JmxTransModule.java#L175). From the stacktrace, it looks like this part is working, but that the pool has been initialized with a `null` factory. The object pool is configured in the same Guice [JmxTransModule](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/guice/JmxTransModule.java#L82-L83) and that's probably close to where the problem is...

**Comment by imochurad** [comment]
Yeah, I figured that there is Guice injection in the mix, the pool object was not null. But the factory field is null. Why is it null? I can't really share with you code atm. It is not in a public repository.

I have main() method that is being executed as an entry point:

`    public static void main(String[] args) throws Exception {
        JmxTransformer transformer = createTransformer();
        transformer.start();
    }

    private static JmxTransformer createTransformer() {
        File f = new File(JMXTRANS_CONFIG_FILE);
        JmxTransConfiguration config = new JmxTransConfiguration();
        config.setJsonFile(f);
        Injector injector = JmxTransModule.createInjector(config);
        JmxTransformer transformer = injector.getInstance(JmxTransformer.class);
        return transformer;
    }
` 
That's pretty much it. json config file I've posted in my previous message. That json file is being generated based on the infrastructure I have running in the cloud. The sample json file posted is being created in local environment though. The issue is being reproduced locally.

**Comment by imochurad** [comment]
I have removed all the sensitive information from my project.
Have a look here, you may disregard Heroku related stuff, in a nutshell, I'm trying to read config from several applications and based on it determine what services may be monitored using JMX.
https://github.com/imochurad/jmxtrans-poc/blob/master/src/main/java/monitoring/MonitoringMain.java
